pkg/sfu/videolayerselector/selectordecisioncache.go

```diff
@@ -0,0 +1,109 @@
+package videolayerselector
+
+import (
+	"fmt"
+)
+
+// ----------------------------------------------------------------------
+
+type selectorDecision int
+
+const (
+	selectorDecisionMissing selectorDecision = iota
+	selectorDecisionDropped
+	selectorDecisionForwarded
+	selectorDecisionUnknown
+)
+
+func (s selectorDecision) String() string {
+	switch s {
+	case selectorDecisionMissing:
+		return "MISSING"
+	case selectorDecisionDropped:
+		return "DROPPED"
+	case selectorDecisionForwarded:
+		return "FORWARDED"
+	case selectorDecisionUnknown:
+		return "UNKNOWN"
+	default:
+		return fmt.Sprintf("%d", int(s))
+	}
+}
+
+// ----------------------------------------------------------------------
+
+type SelectorDecisionCache struct {
+	initialized bool
+	base        uint64
+	last        uint64
+	masks       []uint64
+	numEntries  uint64
+}
+
+func NewSelectorDecisionCache(maxNumElements uint64) *SelectorDecisionCache {
+	numElements := (maxNumElements*2 + 63) / 64
+	return &SelectorDecisionCache{
+		masks:      make([]uint64, numElements),
+		numEntries: numElements * 32, // 2 bits per entry
+	}
+}
+
+func (s *SelectorDecisionCache) AddForwarded(entity uint64) {
+	s.addEntity(entity, selectorDecisionForwarded)
+}
+
+func (s *SelectorDecisionCache) AddDropped(entity uint64) {
+	s.addEntity(entity, selectorDecisionDropped)
+}
+
+func (s *SelectorDecisionCache) GetDecision(entity uint64) (selectorDecision, error) {
+	if !s.initialized || entity > s.last {
+		return selectorDecisionUnknown, nil
+	}
+
+	offset := s.last - entity
+	if offset >= s.numEntries {
+		// asking for something too old
+		return selectorDecisionUnknown, fmt.Errorf("too old, oldest: %d, asking: %d", s.last-s.numEntries+1, entity)
+	}
+
+	return s.getEntity(entity), nil
+}
+
+func (s *SelectorDecisionCache) addEntity(entity uint64, sd selectorDecision) {
+	if !s.initialized {
+		s.initialized = true
+		s.base = entity
```

Comment:
is the first bin guaranteed to arrive in order?

Reply:
Good catch! No, it is not. I will checks to ignore entities less than `base`. Ignoring them is fine.
Also, this needs some more work like I mentioned. When we get a chance to make those changes, can take a look at this again.